### PR TITLE
#162192206 Intercept and handle network errors

### PR DESCRIPTION
**Pull request by mutaimwiti**
**What does this PR do?**

This PR handles network errors application-wide. This is to prevent developers from handling them everywhere that they make API calls.

**Description of Task to be completed?**

Network errors should be handled at one point to prevent repetition of code. Since the entire application uses a single `api` module that wraps `axios` to make http requests, all network errors should be handled by it. A modal should be used to indicate network errors. All areas of the application that have handled network errors should be refactored to be in line with this feature.

**How should this be manually tested?**

This can be tested by following these steps:
1. Open this PRs review app.
2. Click to open any article on the homepage. Allow it to load.
3. Go to your browser dev tools and simulate that the page is offline by checking Offline on the network tab.
4. Wait for 30 seconds and a modal will appear indicating that you're offline. The network error is automatically detected when the application tries to update article reactions and fails.
5. Go back to your browser dev tools disable offline mode on the network tab.
6. Wait for 30 seconds and the network error modal will disappear indicating that there is a connection.

**Any background context you want to provide?**

This solution is making use of `axios` interceptors, https://github.com/axios/axios#interceptors.

**What are the relevant pivotal tracker stories?**

[#162192206](https://www.pivotaltracker.com/story/show/162192206)

**Screenshots**

Simulating network error on browser dev tools.
screen shot 2018-11-27 at 23 52 22

Viewing an article
screen shot 2018-11-29 at 07 09 59

Network error occurs while viewing
screen shot 2018-11-29 at 07 10 39

**Questions**
Anyone that has successfully tested a materialize modal can give ideas on how we can test this.

### CODE REVIEW

File: src/containers/NetworkPopup/index.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import './NetworkPopup.scss';
+
+export class NetWorkPopup extends Component {
+  constructor(props) {
+    super(props);
+    this.ref = React.createRef();
+  }
+
+  onClick = () => {
+    reloadLocation();
```
Comment (paulrimiru) [inline src/containers/NetworkPopup/index.js:15]: Add tests for this function

File: src/containers/NetworkPopup/index.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  };
+
+  getModal = () => this.ref.current;
+
+  componentDidMount() {
+    Materialize.Modal.init(this.getModal(), { dismissible: false });
+  }
+
+  componentDidUpdate() {
+    const { networkError } = this.props;
```
Comment (paulrimiru) [inline src/containers/NetworkPopup/index.js:25]: Add tests for this functionality lines `25 - 30` i feel that they are important to your functionality

File: src/index.js [hunk 3]
```diff
@@ -8,10 +8,14 @@ import * as serviceWorker from './serviceWorker';
 import store from './store';
 
 import './env';
+import ConnectedNetWorkError from "./containers/NetworkPopup";
 
 ReactDOM.render(
   <Provider store={store}>
-    <Router />
+    <>
```
Comment (paulrimiru) [inline src/index.js]: Use `Fragments` instead

**Review by paulrimiru — CHANGES_REQUESTED** [review]

File: src/containers/NetworkPopup/index.js (hunk 1 above)
Comment (mutaimwiti) [reply src/containers/NetworkPopup/index.js:15]: I have added the test @PaulKariukiRimiru

File: src/index.js (hunk 3 above)
Comment (mutaimwiti) [reply src/index.js]: It's done

File: src/containers/NetworkPopup/index.js (hunk 2 above)
Comment (mutaimwiti) [reply src/containers/NetworkPopup/index.js:25]: screen shot 2018-11-29 at 16 35 09

@PaulKariukiRimiru an attempt to test it has proved futile. After a lot research and seeking of peer assistance I have not been able to come up with a test. See https://stackoverflow.com/questions/45833331/jest-can-not-deal-with-insertadjacentelement. This is the result of a `jsdom` shortcoming.

**Review by paulrimiru — APPROVED** [review]

**Pull request merged**